### PR TITLE
Implement client side message validation

### PR DESCRIPTION
**Pull request by minsooerickim**
finished
- [x] invalid questions are still showing up under the questions tab

### CODE REVIEW

File: app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx
Comment (DevelopmentByDavid) [inline app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx]: I would move this to `src/utils/index.ts`.  We may have other places where this could be used in the future.

File: app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx
Comment (DevelopmentByDavid) [inline app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx]: Proper capitalization

File: app/client/src/features/events/Questions/AskQuestion/ValidationError.tsx
Comment (DevelopmentByDavid) [inline app/client/src/features/events/Questions/AskQuestion/ValidationError.tsx]: Hmm, I'm unclear on why this is being made.  I believe a GraphQLError is usually something that's internal to GraphQL, although I could be wrong.  Do you mind explaining the motivation behind making a new Error class and extending GraphQLError?

File: app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx
Comment (DevelopmentByDavid) [inline app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx]: Hmm, this is an important business rule (no questions greater than 1k characters).  I almost think this should be an env variable.  

For example, think about a scenario where someone asks us to update the max question length.  We'd have to remember it's in this file as opposed to just using an env variable.  This comment is more so a note to myself, no changes on your part are required, but feel free to have input on what you think is best.

**Review by DevelopmentByDavid — CHANGES_REQUESTED** [review]
Are you going to be adding validation on the server as follow up?

**Comment by DevelopmentByDavid** [comment]
Realized I forgot to say good work!

**Comment by minsooerickim** [comment]
> Are you going to be adding validation on the server as follow up?

Oh, I for some reason thought the validation had to be done on the client-side. But yes, I can add the validation on the server-side. In that case, would I just get rid of everything on the client side? We would only need to validate the input once, either in the client or the server side, right?

File: app/client/src/features/events/Questions/AskQuestion/ValidationError.tsx
Comment (minsooerickim) [reply app/client/src/features/events/Questions/AskQuestion/ValidationError.tsx]: the main reason why I decided to create that class was because when I tried to just throw a string in AskQuestion.tsx, it gave me an error saying I could only throw an object of a class for some reason. I'm not sure if there's a better way around that

File: app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx
Comment (minsooerickim) [reply app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx]: I totally agree; In that case, I would replace the 1000 with something like MAX_QUESTION_LENGTH and declare that env variable in the .env file; I would also need to declare that env variable in docker if I'm not mistaken, correct?

File: app/client/src/features/events/Questions/AskQuestion/ValidationError.tsx
Comment (DevelopmentByDavid) [reply app/client/src/features/events/Questions/AskQuestion/ValidationError.tsx]: Hmm, when you say throw a string do you mean `throw new 'My Error Message'`?  The normal way to throw an error would be to use the built in one `throw new Error('My Error Message')`. [ Here's some docs on error classes available in the browser. ](https://developer.mozilla.org/en-US/docs/web/javascript/reference/global_objects/error)

File: app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx
Comment (DevelopmentByDavid) [reply app/client/src/features/events/Questions/AskQuestion/AskQuestion.tsx]: Yeah, that would be correct.  However, typically the .env is used for things like the URL to connect to the database or an API key.  I think the .env file could get cluttered if we start putting all of our business rules in there.

Maybe the solution would be to make a file like `rules.ts` where we just export a bunch of hard-coded variables.  So for example `export const QUESTIONS_MAX_LENGTH = 1000`. Then, we can import it like so `import { QUESTIONS_MAX_LENGTH } from '@local/rules';`

**Comment by DevelopmentByDavid** [comment]
> > Are you going to be adding validation on the server as follow up?
> 
> Oh, I for some reason thought the validation had to be done on the client-side. But yes, I can add the validation on the server-side. In that case, would I just get rid of everything on the client side? We would only need to validate the input once, either in the client or the server side, right?

As far as I know (and what I usually do), you do validation on both frontend and backend. (here's a [link](https://stackoverflow.com/questions/37492271/should-i-handle-html-form-validation-on-front-or-backend), but feel free to google some things and let me know if you find anything interesting)

**Comment by DevelopmentByDavid** [comment]
Also, the reason you would do it on both is that theoretically, someone could write some code that sends a request to your backend without ever using your frontend.  That's generally why backend validation is done.  Frontend validation is done so you don't have to ping your server or wait for a request to load before the user gets a response (think of a slow internet connection or someone sending many invalid requests via your client).

**Comment by minsooerickim** [comment]
> Frontend validation is done so you don't have to ping your server or wait for a request to load before the user gets a response (think of a slow internet connection or someone sending many invalid requests via your clie

ah that makes so much more sense. I definitely wasn't thinking of a scenario where someone could directly send a request to the backend. Thank you for your kind explanations!! It's really cool to see, understand, and actually implement some of the good practices when coding like the ones you mentioned above! I'll go ahead and start working on the server-side validation.

File: app/client/src/utils/index.ts
Comment (DevelopmentByDavid) [inline app/client/src/utils/index.ts]: One more question on this, did you make this or find it somewhere?

**Comment by DevelopmentByDavid** [comment]
> > Frontend validation is done so you don't have to ping your server or wait for a request to load before the user gets a response (think of a slow internet connection or someone sending many invalid requests via your clie
> 
> ah that makes so much more sense. I definitely wasn't thinking of a scenario where someone could directly send a request to the backend. Thank you for your kind explanations!! It's really cool to see, understand, and actually implement some of the good practices when coding like the ones you mentioned above! I'll go ahead and start working on the server-side validation.

I'm glad you're enjoying the learning experience!  The server side things can be a separate PR.  It's fine for now if it's only client side.

File: app/client/src/utils/index.ts
Comment (minsooerickim) [reply app/client/src/utils/index.ts]: I found that pattern on one of the stackoverflow pages I believe; Is that something I should avoid?

Comment (DevelopmentByDavid) [reply app/client/src/utils/index.ts]: It's not something bad.  It's usually good to add a source link in the comments.  Also, maybe try testing out the regex on https://regexr.com/.  I think I found the SO post and different people say different things about how good that regex is.

Comment (minsooerickim) [reply app/client/src/utils/index.ts]: I just went ahead and pushed the updated regex. I looked deeper into regex and came across many different ones that people came up with. I found this particular one that passed all of the different possible combinations for a link I could think of. I also added the source link in the comments as you suggested!

File: app/client/src/utils/index.ts
Comment (DevelopmentByDavid) [inline app/client/src/utils/index.ts:21]: Looking at this again, it may be more prudent to try to create a `URL()` and if an error is not thrown, then we know that it's a URL. https://developer.mozilla.org/en-US/docs/Web/API/URL/URL

That's probably not the best way to about using the API, but at the same time, whatever they're using to validate a URL should be much more robust than anything we'd make.

File: app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx
Comment (DevelopmentByDavid) [inline app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx]: Hmm, would this catch if the URL is within a message?  For example, "Please take a look at [some url here]"

My understanding of the current implementation of `isUrl` is that it determines if the entire string is a URL or not.  Whereas the behavior we desire is checking whether or not any substring is a URL.

File: app/client/src/utils/rules.ts
Comment (kadey001) [inline app/client/src/utils/rules.ts]: These could possibly be switched to env variables.

Comment (DevelopmentByDavid) [reply app/client/src/utils/rules.ts]: Ah, I told him to make it this way. https://github.com/c2c-project/prytaneum/pull/259#discussion_r812574922  If you'd prefer env then I'll defer to you.  I'm not the one that has to make sure the gcp and github secrets are updated correctly :)

File: app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx
Comment (minsooerickim) [reply app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx]: right, but in that case we wouldn't be able to use the URL() function to check if it's a valid url. If the string is "hi google.com blah" then putting the whole string into URL() would throw an error. I suppose I could pass in substring word by word separated by whitespace into the URL() function (URL(hi), URL(google.com). However, this approach wouldn't work if the string was something like "higoogle.com". So, wouldn't regex be a more viable option to check for such cases

Comment (minsooerickim) [reply app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx]: isUrl is my function that I wrote with regex that I imported in from index.ts; it does in fact catch the example you mentioned, "Please take a look at [some url here]". I also verified some other cases as well ("higoogle.com"; without the whitespace also gets caught in the function)

Comment (DevelopmentByDavid) [reply app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx]: Okay, seems I was mistaken. Leave as-is.

File: app/client/src/utils/index.ts
Comment (DevelopmentByDavid) [reply app/client/src/utils/index.ts:21]: resolved in other thread

File: app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx
Comment (minsooerickim) [reply app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx]: was there something else you wanted me to do for this issue?

Comment (DevelopmentByDavid) [reply app/client/src/features/events/LiveFeedback/SubmitLiveFeedback.tsx]: Nothing at the moment. I'll give the code a second pass in the next day or two.

**Review by kadey001 — DISMISSED** [review]

**Review by kadey001 — APPROVED** [review]

**Pull request merged**
